### PR TITLE
Statedyn

### PR DESCRIPTION
**Pull request by sanjaylall**
This code makes it easier to add more complex controllers to Callisto. e.g. controllers that switch from one mode to another.

### CODE REVIEW

File: src/opts.jl [hunk 1]
```diff
@@ -197,6 +200,10 @@ function CalOpts(;topology = ("mesh", 3, 2), graph = nothing,
         controller_init = () -> 0.0
     end
 
+    if isnothing(controller_log)
```
Comment (cc10512) [inline src/opts.jl:203]: I wouldn't have guessed that this function implements isnothing, especially given that controller_log gets set to "nothing" in line 119.

Maybe a comment would help.

File: src/statedynamics.jl [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+# asbtract
+
+abstract type StateSystem
+end
+
+##############################################################################
+# PI controller
+
+mutable struct PIStateSystem <: StateSystem
+    x
```
Comment (cc10512) [inline src/statedynamics.jl:17]: It is quite clear what kp and ki represent, but x? Based on the next method, it seems to be the state -- a more explicit name or a comment would help.

File: src/statedynamics.jl [hunk 3]
```diff
@@ -0,0 +1,209 @@
+
+# code for state-space systems
+module StateDynamics
+
+
+
+##############################################################################
+# asbtract
+
+abstract type StateSystem
```
Comment (cc10512) [inline src/statedynamics.jl:10]: For my edification: how come the abstract type does not define the interface?
It looks like all the systems have to define at least a next method that takes the current object/state and the measurement.

**Review by cc10512 — APPROVED** [review]
A couple of minor comments.

File: src/opts.jl (hunk 1 above)
Comment (sanjaylall) [reply src/opts.jl:203]: isnothing is a library function that returns true if it's argument has type Nothing. The controller_log is an unnamed function.

File: src/statedynamics.jl (hunk 2 above)
Comment (sanjaylall) [reply src/statedynamics.jl:17]: Good point.

File: src/statedynamics.jl (hunk 3 above)
Comment (sanjaylall) [reply src/statedynamics.jl:10]: Julia doesn't really have interfaces.

Comment (cc10512) [reply src/statedynamics.jl:10]: then i'm at a loss: what is the point of having abstract types?

Comment (cc10512) [reply src/statedynamics.jl:10]: hmm, from the manual(https://docs.julialang.org/en/v1/manual/types/#man-abstract-types): "Thus, abstract types allow programmers to write generic functions that can later be used as the default method by many combinations of concrete types. Thanks to multiple dispatch, the programmer has full control over whether the default or more specific method is used."
I guess in your case there is no default implementation for next, so there is no point in defining it.

**Pull request merged**